### DISCUSSION TITLE
SF project validate should allow NoTestRun flag

### DISCUSSION THREAD

**Discussion by amarek-cytiva** [forcedotcom/cli#2688] [General]
This discussion is about the possibility of setting the NoTestRun flag during validation if the package does not contain any apex class, triggers, or flows.

Previous commands:
`sfdx force:source:deploy`
allows us to validate the code with the mentioned flag set to NoTestRun for any development environment as well as for Production.

Currently for `sf project deploy validate` the default value is RunLocalTests. We should have the flexibility to decide whether we want to run the unit tests or not during code validation the same way as it was before for sfdx commands.

The workaround for this would be to always run 1 unit test during validation which feels like going back to ant deployments where it was working that way.

Now after the changes, we are forced to run unit tests even on the components that have nothing to do with apex classes like when adding new fields, etc.

Previously setting NoTestRun on the development environment did not execute any unit tests and removing the --test-level flag during validation on Production also did not execute the unit tests execution as long as the package did not contain Apex Class, Apex Trigger, or Flow.

I just found where this issue was described, for the mdapi commands but it was also applied for the source command as well
https://github.com/forcedotcom/cli/issues/1542

To be honest this is a showstopper for using the new version of CLI as we are lacking of the functionality we had before.

**Comment by mcarvin8** [comment]
@amarek-cytiva - I raised a similar issue a while ago. Please see @mshanemc 's response here - https://github.com/forcedotcom/cli/issues/2470#issuecomment-1721450764

`sf project deploy start --dry-run` without tests specified will have Salesforce scan the package and run a dry-run/validation without tests.

**Reply by amarek-cytiva** [reply]
Thank you @mcarvin8 for that,
It is super confusing that we have 2 commands that are very similar to each other.
I do not understand what is the reason to have `sf project deploy validate` and flag `--dry-run` for another command.
It is causing this misunderstanding and overall experience with the cli itself as constantly it needs to have commands updated with completely different values.

I will be testing the same for Production but for developer sandbox it's working as it was for `sfdx force:source:deploy`.

**Reply by mcarvin8** [reply]
I agree it's confusing that there are two commands very similar to each other. I also didn't see anything in the documentation that indicated where `--dry-run` should be used over `validate`.

**Comment by bonzi-123** [comment]
Hello @mdonnalley is there an update on this? I think it would make sense to be able to validate deployments that don't need to execute any test classes using `sf project deploy validate` and `NoTestRun`.

Why do we have to use have `sf project deploy start` and flag `--dry-run` when we're actually performing a validation?

Thanks

Note: I tagged @mdonnalley as I saw you answered in [this thread](https://github.com/forcedotcom/cli/issues/2727) :)
